File: lc_cpp/lcCpp/dp/mostPoints2140.cpp
//
// Created by 24276 on 2025/4/1.
//
#include <vector>
#include <iostream>
using namespace std;

class most_points2140 {
public:
    using ll = long long;
    ll mostPoints(vector<vector<int>>& questions) {
        auto n = questions.size();
     // 用dp存[0,i - 1]能获取得到最大的分数
        vector<ll> dp(n + 1,0);
        //倒着往前算 从前往后计算dp[i]要综合考虑dp[0~i-1]的选择 很麻烦
        for(int i = n - 1;i >= 0;i--) {
            int points = questions[i][0];
            int brainPower = questions[i][1];
            int next = i + brainPower + 1;

            //如果跳过的题目数超出范围 就只能获的当前题目的分数
            if(next > n) {
                dp[i] = std::max(dp[i + 1],(ll)points);
            }else {
                dp[i] = max(dp[i + 1],points + dp[next]);
            }
        }
        return dp[0];
    }
};

int main() {
    most_points2140 solution;
    vector<vector<int>> questions = {
            {3,2},
              {4,3},
                {4,4},
        {2,5},
    };
    cout << solution.mostPoints(questions) <<  endl;
    return 0;
}

File: lc_go/lcGo/dp/mostPoints2140.go
package main

import "fmt"

func mostPoints(questions [][]int) int64 {
	n := len(questions)
	dp := make([]int64, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = 0
	}
	for i := n - 1; i >= 0; i-- {
		points := questions[i][0]
		brainPower := questions[i][1]
		next := i + brainPower + 1

		if next > n { //越界了 只能拿当前的分数
			dp[i] = max(dp[i+1], (int64(points)))
		} else {
			dp[i] = max(dp[i+1], (int64(points))+dp[next])
		}
	}
	return dp[0]
}

func main() {
	questions := [][]int{
		{3, 2},
		{4, 3},
		{4, 4},
		{2, 5},
	}
	res := mostPoints(questions)
	fmt.Println(res)
}
